### ISSUE TITLE
Query :: query source materialized as ValueBuffer rather than entity for some complex cases involving subqueries and conditional operator - causes NRE/InvalidCast

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFramework#4547]
The following query will throw during execution:

```
                var query = from e3 in context.LevelThree
                            join e1 in context.LevelOne
                            on
                                (int?)e3.Id
                            equals
                                (
                                    from subQuery2 in context.LevelTwo
                                    join subQuery3 in context.LevelThree
                                    on
                                        subQuery2 != null ? (int?)subQuery2.Id : null
                                    equals
                                        subQuery3.Level2_Optional_Id
                                    into
                                        grouping
                                    from subQuery3 in grouping.DefaultIfEmpty()
                                    select subQuery3 != null ? (int?)subQuery3.Id : null
                                ).FirstOrDefault()
                            select e1.Id;
```

It is because 

```
subQuery3 != null ? (int?)subQuery3.Id : null
```

required subQuery3 to be materialized as Entity, but instead it is materialized as ValueBuffer.

This is the current query plan for the query:

```
(QueryContext queryContext) => IEnumerable<int> _Select(
    source: IEnumerable<TransparentIdentifier<ValueBuffer, ValueBuffer>> _Join(
        outer: IEnumerable<ValueBuffer> ProjectionQuery(
            queryContext: queryContext, 
            entityType: FunctionalTests.TestModels.ComplexNavigationsModel.Level3
        )
        , 
        inner: IEnumerable<ValueBuffer> ProjectionQuery(
            queryContext: queryContext, 
            entityType: FunctionalTests.TestModels.ComplexNavigationsModel.Level1
        )
        , 
        outerKeySelector: (ValueBuffer e3) => (Nullable<int>) (int) object e3.get_Item(0), 
        innerKeySelector: (ValueBuffer e1) => Nullable<int> FirstOrDefault(
            source: IEnumerable<Nullable<int>> _Select(
                source: IEnumerable<TransparentIdentifier<TransparentIdentifier<Level2, IEnumerable<ValueBuffer>>, ValueBuffer>> _SelectMany(
                    source: IEnumerable<TransparentIdentifier<Level2, IEnumerable<ValueBuffer>>> _GroupJoin(
                        outer: IEnumerable<Level2> EntityQuery(
                            queryContext: queryContext, 
                            entityType: FunctionalTests.TestModels.ComplexNavigationsModel.Level2, 
                            key: Key, 
                            materializer: (IEntityType entityType, ValueBuffer valueBuffer) => 
                            {
                                var var4
                                var4 = new Level2()
                                var4.Id = (int) object valueBuffer.get_Item(0)
                                var4.Level1_Optional_Id = (Nullable<int>) object valueBuffer.get_Item(1)
                                var4.Level1_Required_Id = (int) object valueBuffer.get_Item(2)
                                var4.Name = (string) object valueBuffer.get_Item(3)
                                var4
                            }
                            , 
                            queryStateManager: False
                        )
                        , 
						
						// THIS SHOULD BE MATERIALIZED AS ENTITY, BECAUSE HOW IT IS USED IN THE SELECTOR
                        inner: IEnumerable<ValueBuffer> ProjectionQuery(
                            queryContext: queryContext, 
                            entityType: FunctionalTests.TestModels.ComplexNavigationsModel.Level3
                        )
                        , 
                        outerKeySelector: (Level2 subQuery2) => subQuery2 != null ? (Nullable<int>) subQuery2.Id : null, 
                        innerKeySelector: (ValueBuffer subQuery3) => (Nullable<int>) object subQuery3.get_Item(1), 
                        resultSelector: (Level2 subQuery2, IEnumerable<ValueBuffer> grouping) => TransparentIdentifier<Level2, IEnumerable<ValueBuffer>> CreateTransparentIdentifier(
                            outer: subQuery2, 
                            inner: grouping
                        )
                    )
                    , 
                    collectionSelector: (TransparentIdentifier<Level2, IEnumerable<ValueBuffer>> t0) => IEnumerable<ValueBuffer> DefaultIfEmpty(
                        source: t0.Inner
                    )
                    , 
                    resultSelector: (TransparentIdentifier<Level2, IEnumerable<ValueBuffer>> t0, ValueBuffer subQuery3) => TransparentIdentifier<TransparentIdentifier<Level2, IEnumerable<ValueBuffer>>, ValueBuffer> CreateTransparentIdentifier(
                        outer: t0, 
                        inner: subQuery3
                    )
                )
                , 
				
				// THIS PART IS THE ISSUE: (Level3) object t1.Inner.get_Item(0) != null
                selector: (TransparentIdentifier<TransparentIdentifier<Level2, IEnumerable<ValueBuffer>>, ValueBuffer> t1) => (Level3) object t1.Inner.get_Item(0) != null ? (Nullable<int>) (int) object t1.Inner.get_Item(0) : null
            )
        )
        , 
        resultSelector: (ValueBuffer e3, ValueBuffer e1) => TransparentIdentifier<ValueBuffer, ValueBuffer> CreateTransparentIdentifier(
            outer: e3, 
            inner: e1
        )
    )
    , 
    selector: (TransparentIdentifier<ValueBuffer, ValueBuffer> t0) => (int) object t0.Inner.get_Item(0)
)
```

**Comment by maumar** [comment]
Added test: ComplexNavigationsQueryTestBase -> Query_source_materialization_bug_4547

It repros for InMemory

**Comment by maumar** [comment]
Per conversation with @anpete, assigning to myself

**Comment by maumar** [comment]
Fixed in ff46da5e38f575b45488053d5a6559285af472c9
